Deprecate overwriting existing foreign key constraints

santysisi opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | #7123

This PR introduces a deprecation notice when overwriting an existing foreign key constraint.

### Summary
- Adds a deprecation warning when a foreign key constraint with the same name already exists.
- In Doctrine DBAL 5.0, this behavior will throw an exception instead of silently overwriting the constraint.

santysisi commented:
status: needs work

I think it's better to move that logic into the `_addForeignKeyConstraint` method.

santysisi commented:
status: needs review

I ended up adding the logic to both `_addForeignKeyConstraint` and `addForeignKeyConstraint`.

**Why did I do this?**
The `_addForeignKeyConstraint` method is called by both `addForeignKeyConstraint` and `_construct`. If I only modified `addForeignKeyConstraint`, it wouldn't impact the behavior in `_construct`.

So why modify `addForeignKeyConstraint` as well, if the logic already exists in `_addForeignKeyConstraint`?
That’s because `addForeignKeyConstraint` creates a `ForeignKeyConstraint` with a `name`. For the logic in `_addForeignKeyConstraint` to take effect, the `ForeignKeyConstraint` must not have a `name` set. But since `addForeignKeyConstraint` assigns one, the check in `_addForeignKeyConstraint` would never trigger.

One possible solution could be to remove the `name` generation in `addForeignKeyConstraint`, but I believe that would be a breaking change (BC).

Please let me know if this solution is incorrect or if any changes are needed.
Thanks again! 😄

santysisi commented:
Maybe I'm wrong, but I think the pipeline errors aren't related to this PR

santysisi commented:
should this PR be targeting the 4.3.x branch instead? 🤔

morozov commented:
> Maybe I'm wrong, but I think the pipeline errors aren't related to this PR

Yes, this is unrelated. Sorry about the inconvenience.

> should this PR be targeting the 4.3.x branch instead? 🤔

That sounds right. You can change the base branch of this PR.

santysisi commented:
Base branch changed to `4.3.x` 😄

santysisi commented:
Hey! 👋 I saw this [PR](https://github.com/doctrine/dbal/pull/6972) and went ahead and updated the tests to use `Table::editor()` where appropriate.

morozov commented:
@santysisi, thanks for the PR. Could you let us know whether it was prepared with the help of an AI tool, and if so, to what extent? This will helps us better understand how to approach reviewing it.

santysisi commented:
Hi @morozov, thanks for your comment! Yes, I used an AI tool to help with writing comments (including PHPDoc) and to suggest some method and variable names. This comment is also written with the help of AI, as I'm still working on improving my English, and these tools help me communicate more clearly.

Please let me know if there's anything you'd like me to revise. I'm happy to make changes!

morozov commented on src/Schema/Table.php:
```diff
@@ -502,11 +503,24 @@ public function addForeignKeyConstraint(
+                'fk',
+                $this->_getMaxIdentifierLength(),
+            );
+            if ($this->needsAlternativeForeignKeyName($this->normalizeIdentifier($name), $foreignTableName)) {
```
Looking at the complexity of the solution, I don't think the DBAL should try to accommodate any user input at any cost. The case described in the issue is very theoretical case, while the complexity being added is practical.

Instead, we might detect a collision and emit a deprecation notice (something like "Overwriting an existing foreign key constraint is deprecated"). In 5.0.x, until fixed properly, this will result in an exception.

morozov commented on tests/Schema/TableTest.php:
```diff
@@ -1755,4 +1759,121 @@ public function testInvalidPrimaryKeyIndex(): void
+            )
+            ->create();
+
+        // FK 1: should be added
+        $table->addForeignKeyConstraint('t1', ['t_id'], ['id']);
+        // FK 2: should be added
+        $table->addForeignKeyConstraint('t2', ['t_id'], ['id']);
+        // FK 3: should be added
+        $table->addForeignKeyConstraint('t2', ['t_id'], ['second_id']);
+        // FK 4: duplicate of FK 2 – should NOT be added
+        $table->addForeignKeyConstraint('t2', ['t_id'], ['id']);
+        // FK 5: duplicate of FK 1 – should NOT be added
+        $table->addForeignKeyConstraint('t1', ['t_id'], ['id']);
```
Please keep it to the necessary minimum. It doesn't take 5 constraints to verify a single condition.

Ideally, for each code path we could have a single test with a data provider. The provider would provide some input (e.g. the constraint name (or names), or definition (or definitions) and a boolean that defines whether this input should trigger a deprecation).

The test would apply this input and assert whether the deprecation was or was not triggered. This way, the test will ensure that this is exactly the difference between the inputs (e.g. the presence of a duplicate or its absence) causes triggering a deprecation.

morozov commented on tests/Schema/TableTest.php:
```diff
@@ -1755,4 +1759,121 @@ public function testInvalidPrimaryKeyIndex(): void
+     * @param string[] $expectedReferencingColumns
+     * @param string[] $expectedReferencedColumns
+     */
+    private static function assertForeignKeyMatches(
```
I would try to avoid this code. This is unnecessary boilerplate. If we shift gears and only introduce the deprecation, we will only need to test if the deprecation is triggered w/o actually testing the final state of the object (this is already done in other existing tests).

santysisi commented:
Hi @morozov, thanks a lot for taking the time to review the PR, I really appreciate it! 😊

If I understood correctly, the idea now is to trigger a deprecation notice only in case of a collision, but still continue with the current behavior (i.e., overwrite).
Could you please confirm that? 🙏

If so, I can update the PR this weekend to reflect that behavior.

Thanks again! ❤️

morozov commented:
> the idea now is to trigger a deprecation notice only in case of a collision, but still continue with the current behavior (i.e., overwrite).

Correct. This will become an exception in 5.0.x.

santysisi committed:
Deprecate overwriting existing FK constraints
Overwriting an existing foreign key constraint is now deprecated
This behavior will no longer be allowed in Doctrine DBAL 5.0, where
an exception will be thrown instead of silently overwriting the
constraint.

santysisi commented:
Hi! 😄
I've updated the PR to trigger a deprecation. Additionally, I've changed the base branch, as well as the PR title and description.

morozov commented on src/Schema/Table.php:
```diff
@@ -938,6 +939,14 @@ protected function _addForeignKeyConstraint(ForeignKeyConstraint $constraint): s
             );
 
         $name = $this->normalizeIdentifier($name);
+        if (array_key_exists($name, $this->_fkConstraints)) {
```
I'd replace `array_key_exists()` with `isset()` since it's historically preferable and we don't expect null values.

morozov commented on tests/Schema/TableTest.php:
```diff
@@ -1768,4 +1768,22 @@ public function testInvalidPrimaryKeyIndex(): void
+            )
+            ->create();
+
+        $table->addForeignKeyConstraint('bar', ['id'], ['id'], ['foo' => 'bar']);
```
`['foo' => 'bar']` is not a valid set of options, and they shouldn't be necessary for the test. Please remove.

morozov commented on UPGRADE.md:
```diff
@@ -8,6 +8,10 @@ awareness about deprecated code.
 
+## Deprecated overwriting foreign key constraints
+
+Overwriting an existing foreign key constraint has been deprecated.
```
Let's try to be more specific. How about this:

> Adding a foreign key constraint with a name that matches an existing one, whether explicitly specified or auto-generated, has been deprecated".

morozov reviewed (commented):
We're getting close. Besides a few more comments, looks good.

santysisi commented:
Thanks a lot for your suggestions, the changes have been made

derrabus commented on UPGRADE.md:
```diff
@@ -8,6 +8,11 @@ awareness about deprecated code.
+## Deprecated overwriting foreign key constraints
+
+Adding a foreign key constraint with a name that matches an existing one, whether explicitly specified or
+auto-generated, has been deprecated.
```
What should I do instead?

morozov replied:
Good question. In order to replace an existing constraint, drop it first via `dropForeignKey()`.

@santysisi please add this to the upgrade notes.

morozov commented:
Thanks, @santysisi. Congrats on your first contribution!

santysisi commented:
Thanks for your help! 😄

morozov commented:
@santysisi, I merged the changes from this and your other PR up into `5.0.x`. If you're up for it, you can convert these deprecations to exceptions.

santysisi commented:
Yep! 😄
I'll be opening those PRs today ❤️

Edit: I don't have enough time today 😔, I will try to do it this weekend

santysisi commented:
Hi @morozov,

While working on the new PR (to replace the deprecation with throw and error), I encountered an issue (reproducible with SQLite — I’ve tested with SQLite and PostgreSQL only so far) in this method [getForeignKeysInAlteredTable](https://github.com/doctrine/dbal/blob/5.0.x/src/Platforms/SQLitePlatform.php#L860C22-L860C50).

The problem stems from this change: https://github.com/doctrine/dbal/pull/7116, specifically this part:
https://github.com/doctrine/dbal/blob/5.0.x/src/Schema/Table.php#L707-L710

Previously, we used [unset($foreignKeys[$constraintName])](https://github.com/doctrine/dbal/blob/5.0.x/src/Platforms/SQLitePlatform.php#L908), but since the method now returns a list, unsetting by key doesn’t work anymore.

Before proceeding with the final deprecation removal and introducing the new error, I think we should address this issue first.

I’d be happy to take a shot at fixing it! 😄

I see two possible approaches:

Adjust the existing method:
Keep the current method that returns a list, but calculate the index of the constraint in the list and unset() by position. Alternatively, rebuild the array to use the constraint names as keys.

Introduce a new method:
Add a separate method that returns the foreign keys as an associative array (like before), without modifying the current list-based method.

Let me know what you think!

santysisi commented:
Just as an example — if we don’t want to introduce a new method that returns an associative array with constraint names as keys, we could handle it like this:

```php
$foreignKeys = [];
foreach ($oldTable->getForeignKeys() as $key => $foreignKey) {
    $constraintName = $foreignKey->getObjectName();
    
    if ($constraintName === null) {
        $foreignKeys[$key] = $foreignKey;
        continue;
    }

    $foreignKeys[strtolower($constraintName->getIdentifier()->getValue())] = $foreignKey;
}
```
This way, we preserve the ability to reference foreign keys by their constraint names (like before), while still using the current method that returns a list.

It’s a simple alternative to avoid breaking changes or adding a new method, instead of just doing:
```php
$foreignKeys = $oldTable->getForeignKeys();
```


Another alternative could be a deeper refactor 🤔

morozov commented:
> Previously, we used [unset($foreignKeys[$constraintName])](https://github.com/doctrine/dbal/blob/5.0.x/src/Platforms/SQLitePlatform.php#L908), but since the method now returns a list, unsetting by key doesn’t work anymore.

Yeah, I noticed this issue as well. Indexing by name makes sense. The only thing is that your code still mixes name-based and positional keys, which is prone to conflicts (`0` could be both a position of an unnamed constraint and a name of a named one).

Instead, you could introduce and populate a map of constraint names to their numeric keys in the `$foreignKeys` array (e.g. `$keysByName`) – only for those constraints that have a name. Then, while processing dropped FKs:
1. Assert that it has a name
2. Check if there's an entry in `$keysByName` for its name
3. If there is, unset the corresponding key in `$foreignKeys`
